File: src/main/java/recipeapplication/service/RecipeService.java
package recipeapplication.service;

import org.hibernate.Hibernate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import recipeapplication.dto.*;
import recipeapplication.exception.IngredientDoesNotExistException;
import recipeapplication.exception.RecipeDoesNotExistException;
import recipeapplication.exception.SameUsernameException;
import recipeapplication.model.*;
import recipeapplication.repository.IngredientRepository;
import recipeapplication.repository.RecentlyViewedRepository;
import recipeapplication.repository.RecipeRepository;
import recipeapplication.utility.RecipeTime;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import javax.persistence.EntityManager;
import javax.transaction.Transactional;

@Service
@Transactional
public class RecipeService
{
    private RecipeRepository recipeRepository;
    private IngredientRepository ingredientRepository;
    private AuthService authService;
    private RecentlyViewedRepository recentlyViewedRepository;
    private EntityManager entityManager;

    @Autowired
    public RecipeService(
            RecipeRepository recipeRepository,
            IngredientRepository ingredientRepository,
            AuthService authService,
            RecentlyViewedRepository recentlyViewedRepository,
            EntityManager entityManager)
    {
        this.recipeRepository = recipeRepository;
        this.ingredientRepository = ingredientRepository;
        this.authService = authService;
        this.recentlyViewedRepository = recentlyViewedRepository;
        this.entityManager = entityManager;
    }

    public Recipe createRecipe(CreateRecipeDto createRecipeDto)
    {
        User user = authService.getLoggedInUser();

        Recipe recipe = new Recipe();

        recipe.setTitle(createRecipeDto.getTitle());
        recipe.setUser(user);
        recipe.setRating(0L);
        recipe.setServes(1L);
        recipe.setDifficulty("Medium");
        recipe.setTotalTime("00:00");
        recipe.setCookTime("00:00");
        recipe.setPrepTime("00:00");

        return recipeRepository.save(recipe);
    }

    public List<Recipe> getRecipes()
    {
        User user = authService.getLoggedInUser();

        return recipeRepository.findByUser(user);
    }

    public Recipe getRecipe(Long id) throws RecipeDoesNotExistException
    {
        Optional<Recipe> result = recipeRepository.findByIdAndUser(id, authService.getLoggedInUser());

        if (!result.isPresent())
        {
            throw new RecipeDoesNotExistException();
        }

        return result.get();
    }

    public void deleteRecipe(RecipeDto recipeDto) throws RecipeDoesNotExistException
    {
        getRecipe(recipeDto.getId());

        recipeRepository.deleteById(recipeDto.getId());
    }

    public void updateRecipe(RecipeDto recipeDto) throws RecipeDoesNotExistException
    {
        Recipe recipe = getRecipe(recipeDto.getId());

        recipe.setNotes(recipeDto.getNotes());
        recipe.setRating(recipeDto.getRating());
        recipe.setServes(recipeDto.getServes());
        recipe.setCookTime(recipeDto.getCookTime());
        recipe.setPrepTime(recipeDto.getPrepTime());
        recipe.setTotalTime(RecipeTime.combineCookAndPrepTime(recipeDto.getCookTime(), recipeDto.getPrepTime()));
        recipe.setDifficulty(recipeDto.getDifficulty());

        recipeRepository.save(recipe);
    }

    public void shareRecipe(RecipeDto recipeDto, User user) throws RecipeDoesNotExistException, SameUsernameException
    {
        User loggedInUser = authService.getLoggedInUser();
        if (loggedInUser.getUsername().equals(user.getUsername()))
        {
            throw new SameUsernameException();
        }

        Recipe recipe = getRecipe(recipeDto.getId());

        // Clone the existing recipe
        Hibernate.initialize(recipe.getIngredients());
        Hibernate.initialize(recipe.getSteps());
        entityManager.detach(recipe);
        recipe.setId(null);

        recipe.setSharedBy(authService.getLoggedInUser().getUsername());
        recipe.setUser(user);

        List<Ingredient> ingredients = new ArrayList<>();
        for (Ingredient ingredient : recipe.getIngredients())
        {
            ingredients.add(
                    new Ingredient(ingredient.getDescription(), ingredient.getMetric(), user)
            );
        }

        List<Step> steps = new ArrayList<>();
        for (Step step : recipe.getSteps())
        {
            steps.add(
                    new Step(step.getDescription())
            );
        }

        recipe.setIngredients(ingredients);
        recipe.setSteps(steps);

        recipeRepository.save(recipe);
    }

    public void addRecentlyViewed(Recipe recipe)
    {
        Recipe mostRecentRecipe = null;
        List<RecentlyViewed> recentlyViewedRecipes = getRecentlyViewed();

        if (recentlyViewedRecipes.size() > 0)
        {
            mostRecentRecipe = recentlyViewedRecipes.get(0).getRecipe();
        }

        if (mostRecentRecipe == null || !mostRecentRecipe.getId().equals(recipe.getId()))
        {
            RecentlyViewed recentlyViewed = new RecentlyViewed();
            recentlyViewed.setRecipe(recipe);
            recentlyViewed.setUser(recipe.getUser());
            recentlyViewedRepository.save(recentlyViewed);
        }
    }

    public List<RecentlyViewed> getRecentlyViewed()
    {
        User user = authService.getLoggedInUser();

        return recentlyViewedRepository.findTop5ByUserOrderByIdDesc(user);
    }

    public void deleteAllRecipes()
    {
        User user = authService.getLoggedInUser();

        recipeRepository.deleteAllByUser(user);
    }

    public Ingredient addIngredient(IngredientDto ingredientDto) throws RecipeDoesNotExistException
    {
        Recipe recipe = getRecipe(ingredientDto.getRecipe());
        User user = authService.getLoggedInUser();

        Ingredient ingredient = new Ingredient(ingredientDto.getDescription(), ingredientDto.getQuantity(), user);

        recipe.addIngredient(ingredient);

        recipeRepository.save(recipe);

        // Simply returning ingredient does not set the ID required by the front end
        return recipe.getIngredients().get(
                recipe.getIngredients().size() - 1
        );
    }

    public Step addStep(CreateStepDto stepDto) throws RecipeDoesNotExistException
    {
        Recipe recipe = getRecipe(stepDto.getRecipe());

        Step step = new Step(stepDto.getDescription());

        recipe.addStep(step);

        recipeRepository.save(recipe);

        // Simply returning the step does not set the ID required by the front end
        return recipe.getSteps().get(
                recipe.getSteps().size() - 1
        );
    }

    public void deleteIngredient(DeleteIngredientDto ingredientDto) throws RecipeDoesNotExistException, IngredientDoesNotExistException
    {
        User user = authService.getLoggedInUser();
        Recipe recipe = getRecipe(ingredientDto.getRecipeId());
        Optional<Ingredient> ingredient = ingredientRepository.findByIdAndUser(ingredientDto.getIngredientId(), user);

        if (!ingredient.isPresent())
        {
            throw new IngredientDoesNotExistException();
        }

        for (Ingredient recipeIngredient : recipe.getIngredients())
        {
            if (recipeIngredient == ingredient.get())
            {
                recipe.getIngredients().remove(recipeIngredient);
                recipeRepository.save(recipe);
                break;
            }
        }
    }

    public void deleteStep(DeleteStepDto stepDto) throws RecipeDoesNotExistException
    {
        Recipe recipe = getRecipe(stepDto.getRecipeId());
        Step step = extractStepFromRecipe(stepDto.getStepId(), recipe);

        if (step != null)
        {
            recipe.getSteps().remove(step);
            recipeRepository.save(recipe);
        }
    }

    public void updateStep(UpdateStepDto stepDto) throws RecipeDoesNotExistException
    {
        Recipe recipe = getRecipe(stepDto.getRecipe());
        Step step = extractStepFromRecipe(stepDto.getId(), recipe);

        if (step != null)
        {
            step.setDescription(stepDto.getDescription());
            recipeRepository.save(recipe);
        }
    }

    private Step extractStepFromRecipe(Long stepId, Recipe recipe)
    {
        for (Step step : recipe.getSteps())
        {
            if (step.getId().equals(stepId))
            {
                return step;
            }
        }

        return null;
    }
}

File: src/main/kotlin/recipeapplication/service/RecipeService.kt
package recipeapplication.service

import org.hibernate.Hibernate
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import recipeapplication.dto.*
import recipeapplication.exception.IngredientDoesNotExistException
import recipeapplication.exception.RecipeDoesNotExistException
import recipeapplication.exception.SameUsernameException
import recipeapplication.model.*
import recipeapplication.repository.IngredientRepository
import recipeapplication.repository.RecentlyViewedRepository
import recipeapplication.repository.RecipeRepository
import recipeapplication.utility.combineCookAndPrepTime
import javax.persistence.EntityManager
import javax.transaction.Transactional

@Service
@Transactional
open class RecipeService @Autowired constructor(
        private val recipeRepository: RecipeRepository,
        private val ingredientRepository: IngredientRepository,
        private val authService: AuthService,
        private val recentlyViewedRepository: RecentlyViewedRepository,
        private val entityManager: EntityManager
) {
    companion object {
        private const val DEFAULT_DIFFICULTY = "Medium"
        private const val DEFAULT_TIME = "00:00"
        private const val DEFAULT_RATING = 0L
        private const val DEFAULT_SERVES = 1L
    }

    fun createRecipe(createRecipeDto: CreateRecipeDto): Recipe {
        return recipeRepository.save(
                Recipe().apply {
                    title = createRecipeDto.title
                    user = authService.loggedInUser
                    rating = DEFAULT_RATING
                    serves = DEFAULT_SERVES
                    difficulty = DEFAULT_DIFFICULTY
                    totalTime = DEFAULT_TIME
                    cookTime = DEFAULT_TIME
                    prepTime = DEFAULT_TIME
                }
        )
    }

    val recipes: List<Recipe>
        get() {
            return recipeRepository.findByUser(
                    authService.loggedInUser
            )
        }

    @Throws(RecipeDoesNotExistException::class)
    fun getRecipe(id: Long): Recipe {
        val result = recipeRepository.findByIdAndUser(id, authService.loggedInUser)

        if (!result.isPresent) {
            throw RecipeDoesNotExistException()
        }

        return result.get()
    }

    @Throws(RecipeDoesNotExistException::class)
    fun deleteRecipe(recipeDto: RecipeDto) {
        getRecipe(recipeDto.id)
        recipeRepository.deleteById(recipeDto.id)
    }

    @Throws(RecipeDoesNotExistException::class)
    fun updateRecipe(recipeDto: RecipeDto) {
        recipeRepository.save(
                getRecipe(recipeDto.id).apply {
                    notes = recipeDto.notes
                    rating = recipeDto.rating
                    serves = recipeDto.serves
                    cookTime = recipeDto.cookTime
                    prepTime = recipeDto.prepTime
                    totalTime = combineCookAndPrepTime(recipeDto.cookTime, recipeDto.prepTime)
                    difficulty = recipeDto.difficulty
                }
        )
    }

    @Throws(RecipeDoesNotExistException::class, SameUsernameException::class)
    fun shareRecipe(recipeDto: RecipeDto, user: User) {
        if (authService.loggedInUser.username == user.username) {
            throw SameUsernameException()
        }

        val recipe = getRecipe(recipeDto.id)

        // Clone the existing recipe
        Hibernate.initialize(recipe.ingredients)
        Hibernate.initialize(recipe.steps)
        entityManager.detach(recipe)

        recipe.apply {
            id = null
            sharedBy = authService.loggedInUser.username
            setUser(user)
            ingredients = recipe.ingredients.map { ingredient -> Ingredient(ingredient.description, ingredient.metric, user) }
            steps = recipe.steps.map { step -> Step(step.description) }
        }

        recipeRepository.save(recipe)
    }

    fun addRecentlyViewed(recipe: Recipe) {
        if (recentlyViewed.isEmpty() || recentlyViewed[0].recipe.id != recipe.id) {
            recentlyViewedRepository.save(
                    RecentlyViewed().apply {
                        setRecipe(recipe)
                        user = recipe.user
                    }
            )
        }
    }

    val recentlyViewed: List<RecentlyViewed>
        get() {
            return recentlyViewedRepository.findTop5ByUserOrderByIdDesc(
                    authService.loggedInUser
            )
        }

    fun deleteAllRecipes() {
        recipeRepository.deleteAllByUser(
                authService.loggedInUser
        )
    }

    @Throws(RecipeDoesNotExistException::class)
    fun addIngredient(ingredientDto: IngredientDto): Ingredient {
        val recipe = getRecipe(ingredientDto.recipe).apply {
            addIngredient(
                    Ingredient(ingredientDto.description, ingredientDto.quantity, authService.loggedInUser)
            )
        }

        recipeRepository.save(recipe)

        // Simply returning ingredient does not set the ID required by the front end
        return recipe.ingredients.last()
    }

    @Throws(RecipeDoesNotExistException::class)
    fun addStep(stepDto: CreateStepDto): Step {
        val recipe = getRecipe(stepDto.recipe).apply {
            addStep(
                    Step(stepDto.description)
            )
        }

        recipeRepository.save(recipe)

        // Simply returning the step does not set the ID required by the front end
        return recipe.steps.last()
    }

    @Throws(RecipeDoesNotExistException::class, IngredientDoesNotExistException::class)
    fun deleteIngredient(ingredientDto: DeleteIngredientDto) {
        val recipe = getRecipe(ingredientDto.recipeId)
        val ingredient = ingredientRepository.findByIdAndUser(ingredientDto.ingredientId, authService.loggedInUser)

        if (!ingredient.isPresent) { // TODO can Kotlin remove the need for Optional?
            throw IngredientDoesNotExistException()
        }

        if (recipe.ingredients.remove(ingredient.get())) {
            recipeRepository.save(recipe)
        }
    }

    @Throws(RecipeDoesNotExistException::class)
    fun deleteStep(stepDto: DeleteStepDto) {
        val recipe = getRecipe(stepDto.recipeId)
        val step = recipe.steps.firstOrNull { step -> step.id == stepDto.stepId }

        if (step != null) { // TODO sounds like this should throw an exception rather than be null safe
            recipe.steps.remove(step)
            recipeRepository.save(recipe)
        }
    }

    @Throws(RecipeDoesNotExistException::class)
    fun updateStep(stepDto: UpdateStepDto) {
        val recipe = getRecipe(stepDto.recipe)
        val step = recipe.steps.firstOrNull { step -> step.id == stepDto.id }

        if (step != null) { // TODO sounds like this should throw an exception rather than be null safe
            step.description = stepDto.description
            recipeRepository.save(recipe)
        }
    }
}